File: 1073-number-of-enclaves/number-of-enclaves.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<int>>& grid, int drow[], int dcol[]) {
        int m = grid.size();
        int n = grid[0].size();
        grid[row][col] = -1; 
        for (int i = 0; i < 4; i++) {
            int nrow = row + drow[i];
            int ncol = col + dcol[i];
            if (nrow >= 0 && nrow < m && ncol >= 0 && ncol < n && grid[nrow][ncol] == 1) {
                dfs(nrow, ncol, grid, drow, dcol);
            }
        }
    }

    int numEnclaves(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int drow[] = {-1, 0, +1, 0};
        int dcol[] = {0, +1, 0, -1};

        for (int i = 0; i < m; i++) {
            if (grid[i][0] == 1) dfs(i, 0, grid, drow, dcol);       
            if (grid[i][n-1] == 1) dfs(i, n-1, grid, drow, dcol);   
        }
        for (int i = 0; i < n; i++) {
            if (grid[0][i] == 1) dfs(0, i, grid, drow, dcol);       
            if (grid[m-1][i] == 1) dfs(m-1, i, grid, drow, dcol);   
        }

        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    count++;
                }
            }
        }
        return count;
    }
};

File: 1073-number-of-enclaves/number-of-enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        m=len(grid)
        n=len(grid[0])
        def dfs(i,j):
            grid[i][j]=0
            dir=[[0,1],[1,0],[0,-1],[-1,0]]

            for dr in dir:
                newr=i+dr[0]
                newc=j+dr[1]

                if newr>=0 and newr<m and newc>=0 and newc<n and grid[newr][newc]==1:
                    dfs(newr,newc)

        for i in range(m):
            for j in range(n):
                if i==0 or i==m-1 or j==0 or j==n-1:
                    if grid[i][j]==1:
                        dfs(i,j)
                    

        cnt=0
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    cnt+=1
        return cnt